jnimarshalmethod-gen.exe Windows support

radekdoulik opened the pull request:
Added `monow.exe` to the `host-mxe-*` directories and to the bundle.

radekdoulik commented:
> Should this PR also include an update to the _GenerateJniMarshalMethods target so that jnimarshalmethod-gen.exe is executed on Windows?

It should (and does now :-). It has also *do-not-merge* label, because I am still working on it.

radekdoulik commented:
I have got to the point, where we can now run the `jnimarshalmethod-gen.exe` on Windows. It crashes though:

```
( set MONO_PATH=C:\Program Files\Microsoft Visual Studio\2017\Enterprise\Common7\IDE\ReferenceAssemblies\Microsoft\Framework\MonoAndroid\v1.0\;obj\Release\81\linksrc\ ) & "C:\Program Files\Microsoft Visual Studio\2017\Enterprise\MSBuild\Xamarin\Andro
  id\lib\host-mxe-Win32\monow.exe" --debug "C:\Program Files\Microsoft Visual Studio\2017\Enterprise\MSBuild\Xamarin\Android\jnimarshalmethod-gen.exe" --jvm="C:\Program Files\Java\jdk1.8.0_172\jre\bin\client\jvm.dll"  "obj\Release\81\linksrc\MobileApp2
  .Android.dll" "obj\Release\81\linksrc\FormsViewGroup.dll" "obj\Release\81\linksrc\MobileApp2.dll" "obj\Release\81\linksrc\Xamarin.Android.Arch.Core.Common.dll" "obj\Release\81\linksrc\Xamarin.Android.Arch.Lifecycle.Common.dll" "obj\Release\81\linksrc
  \Xamarin.Android.Arch.Lifecycle.Runtime.dll" "obj\Release\81\linksrc\Xamarin.Android.Support.Animated.Vector.Drawable.dll" "obj\Release\81\linksrc\Xamarin.Android.Support.Annotations.dll" "obj\Release\81\linksrc\Xamarin.Android.Support.Compat.dll" "o
  bj\Release\81\linksrc\Xamarin.Android.Support.Core.UI.dll" "obj\Release\81\linksrc\Xamarin.Android.Support.Core.Utils.dll" "obj\Release\81\linksrc\Xamarin.Android.Support.Design.dll" "obj\Release\81\linksrc\Xamarin.Android.Support.Fragment.dll" "obj\
  Release\81\linksrc\Xamarin.Android.Support.Media.Compat.dll" "obj\Release\81\linksrc\Xamarin.Android.Support.Transition.dll" "obj\Release\81\linksrc\Xamarin.Android.Support.v4.dll" "obj\Release\81\linksrc\Xamarin.Android.Support.v7.AppCompat.dll" "ob
  j\Release\81\linksrc\Xamarin.Android.Support.v7.CardView.dll" "obj\Release\81\linksrc\Xamarin.Android.Support.v7.MediaRouter.dll" "obj\Release\81\linksrc\Xamarin.Android.Support.v7.Palette.dll" "obj\Release\81\linksrc\Xamarin.Android.Support.v7.Recyc
  lerView.dll" "obj\Release\81\linksrc\Xamarin.Android.Support.Vector.Drawable.dll" "obj\Release\81\linksrc\Xamarin.Forms.Core.dll" "obj\Release\81\linksrc\Xamarin.Forms.Platform.Android.dll" "obj\Release\81\linksrc\Xamarin.Forms.Platform.dll" "obj\Rel
  ease\81\linksrc\Xamarin.Forms.Xaml.dll" "obj\Release\81\linksrc\Mono.Android.dll" (TaskId:242)
   (TaskId:242)
  Unhandled Exception: (TaskId:242)
  System.NotSupportedException: The given path's format is not supported. (TaskId:242)
   (TaskId:242)
    at System.AppDomainSetup.GetAppBase (System.String appBase) [0x000e2] in /Users/rodo/git/xa5/external/mono/mcs/class/corlib/System/AppDomainSetup.cs:166  (TaskId:242)
   (TaskId:242)
    at System.AppDomainSetup.get_ApplicationBase () [0x00000] in /Users/rodo/git/xa5/external/mono/mcs/class/corlib/System/AppDomainSetup.cs:170  (TaskId:242)
   (TaskId:242)
    at System.AppDomain.CreateDomain (System.String friendlyName, System.Security.Policy.Evidence securityInfo, System.AppDomainSetup info) [0x00046] in /Users/rodo/git/xa5/external/mono/mcs/class/corlib/System/AppDomain.cs:1022  (TaskId:242)
   (TaskId:242)
    at System.AppDomain.CreateDomain (System.String friendlyName) [0x00000] in /Users/rodo/git/xa5/external/mono/mcs/class/corlib/System/AppDomain.cs:989  (TaskId:242)
   (TaskId:242)
    at Xamarin.Android.Tools.JniMarshalMethodGenerator.App.Main (System.String[] args) [0x00001] in <a99a4146dcfa43c080015e4eb31a1ea2>:0  (TaskId:242)
  [ERROR] FATAL UNHANDLED EXCEPTION: System.NotSupportedException: The given path's format is not supported. (TaskId:242)
   (TaskId:242)
    at System.AppDomainSetup.GetAppBase (System.String appBase) [0x000e2] in /Users/rodo/git/xa5/external/mono/mcs/class/corlib/System/AppDomainSetup.cs:166  (TaskId:242)
   (TaskId:242)
    at System.AppDomainSetup.get_ApplicationBase () [0x00000] in /Users/rodo/git/xa5/external/mono/mcs/class/corlib/System/AppDomainSetup.cs:170  (TaskId:242)
   (TaskId:242)
    at System.AppDomain.CreateDomain (System.String friendlyName, System.Security.Policy.Evidence securityInfo, System.AppDomainSetup info) [0x00046] in /Users/rodo/git/xa5/external/mono/mcs/class/corlib/System/AppDomain.cs:1022  (TaskId:242)
   (TaskId:242)
    at System.AppDomain.CreateDomain (System.String friendlyName) [0x00000] in /Users/rodo/git/xa5/external/mono/mcs/class/corlib/System/AppDomain.cs:989  (TaskId:242)
   (TaskId:242)
    at Xamarin.Android.Tools.JniMarshalMethodGenerator.App.Main (System.String[] args) [0x00001] in <a99a4146dcfa43c080015e4eb31a1ea2>:0  (TaskId:242)
```
So now I need to fix the `jnimarshalmethod-gen.exe` itself.

jonathanpeppers commented on src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets:
```diff
@@ -2151,8 +2151,17 @@ because xbuild doesn't support framework reference assemblies.
   <Exec
-       Command="MONO_PATH=&quot;$(_XATargetFrameworkDirectories):$(MonoAndroidLinkerInputDir)&quot; &quot;$(MonoAndroidBinDirectory)\mono&quot; --debug &quot;$(MonoAndroidToolsDirectory)\jnimarshalmethod-gen.exe&quot; --jvm=&quot;$(JdkJvmPath)&quot; $(AndroidGenerateJniMarshalMethodsAdditionalArguments) @(_AssembliesToProcess->'&quot;$(MonoAndroidLinkerInputDir)%(Filename)%(Extension)&quot;', ' ' ...
+       Command="$(_MonoPath) &quot;$(_MonoBinary)&quot; --debug &quot;$(MonoAndroidToolsDirectory)\jnimarshalmethod-gen.exe&quot; --jvm=&quot;$(JdkJvmPath)&quot; $(AndroidGenerateJniMarshalMethodsAdditionalArguments) @(_AssembliesToProcess->'&quot;$(MonoAndroidLinkerInputDir)%(Filename)%(Extension)&quot;', ' ')"
   />
```
Just a thought, but you could make a new `JniMarshalMethodGen` MSBuild task that inherits from [ToolTask](https://docs.microsoft.com/en-us/dotnet/api/microsoft.build.utilities.tooltask?view=netframework-4.7.2). It might a lot cleaner to do all this logic in C# that checks for Windows, 32-bit, 64-bit, etc. There is also an `EnvironmentVariables` property for the new process, where you could set `MONO_PATH`.

radekdoulik replied:
Yes, it was flowing in that direction for some time. Today I decided to finally do it, so that I can try different quotation on Windows to see whether it would help with the above issue.

It didn't help, but at least we have the task now, and so we have the OS specific code hidden in the task from `targets` file.

radekdoulik commented:
I tracked the unhandled exception on Windows to an issue of `System.IO.Path.GetFullPath` in mono. Where we get invalid path like `C:\C:\Program Files\...`

It looks like the `mscorlib` comes from our Android assemblies. So I guess I need to make it use the host one.

radekdoulik commented:
With mscorlib + Facades from mono/Windows installation I am able to get further along. They come from `C:\Program Files\Mono\lib\mono\4.5` location.

@luhenry, Looks like we will need bcl assemblies for windows in the mono SDK for Android, or at least part of them with above mentioned dll's. Would it be possible to have these available as part of https://github.com/xamarin/xamarin-android/pull/2253 bump?

radekdoulik commented:
I have now run into another issue on windows.
```
Error: jnimarshalmethod-gen: Unable to create Java VM
System.Exception: Could not load JVM path `C:\Program Files\Java\jdk1.8.0_172\jre\bin\client\jvm.dll` (-1002)!
  at Java.Interop.JreRuntime.CreateJreVM (Java.Interop.JreRuntimeOptions builder) [0x00075] in <37e1efe7c6864901b006814060f7b61a>:0
  at Java.Interop.JreRuntime..ctor (Java.Interop.JreRuntimeOptions builder) [0x00000] in <37e1efe7c6864901b006814060f7b61a>:0
  at Java.Interop.JreRuntimeOptions.CreateJreVM () [0x00001] in <37e1efe7c6864901b006814060f7b61a>:0
  at Xamarin.Android.Tools.JniMarshalMethodGenerator.App.CreateJavaVM (System.String jvmDllPath) [0x00010] in <a7f7c2545bab42409f04383f1ee3de40>:0
```
```
dir "C:\Program Files\Java\jdk1.8.0_172\jre\bin\client\jvm.dll"
 
 Directory of C:\Program Files\Java\jdk1.8.0_172\jre\bin\client

11/21/2018  05:32 PM         3,885,480 jvm.dll
```
That comes from https://github.com/xamarin/java.interop/blob/2e682251c3942f2323ec39b21830b39fd74bbf67/src/java-interop/java-interop-jvm.c#L32-L37

So it looks like we might need to use `LoadLibrary` call on Windows instead of `dlopen`.

jonpryor commented:
@radekdoulik wrote:

> So it looks like we might need to use `LoadLibrary` call on Windows instead of `dlopen`.

This doesn't make sense to me, as `libmono-android.debug.dll` on Windows *should* be using [`external`](https://github.com/xamarin/xamarin-android/tree/master/external)`/`[`dlfcn-win32`](https://github.com/dlfcn-win32/dlfcn-win32/tree/ef7e412d3fa2b511e660342e237536c235e51ee4), which implements `dlopen()` and *does* use `LoadLibrary()` (`LoadLibraryEx()`, specifically):

https://github.com/dlfcn-win32/dlfcn-win32/blob/ef7e412d3fa2b511e660342e237536c235e51ee4/dlfcn.c#L283-L284

So either we're *not* actually using `dlfcn-win32` -- in which case...how does anything work in general?  Shouldn't we get an "unresolved symbol error" when trying to resolve `dlopen` if we're not using `dlfcn-win32`? -- or we *are*, and...something else is wrong?

Very confusing, regardless.

radekdoulik commented:
`jvm.dll` dependencies (so that I don't have to look it up again):
```
dumpbin.exe" /dependents "C:\Program Files\Java\jdk1.8.0_172\jre\bin\client\jvm.dll"
...
  Image has the following dependencies:

    KERNEL32.dll
    USER32.dll
    ADVAPI32.dll
    WSOCK32.dll
    WINMM.dll
    VERSION.dll
    PSAPI.DLL
    MSVCR100.dll
...
```

radekdoulik commented:
My guess is `MSVCR100.dll` was missing. It is located in `jdk1.8.0_172\jre\bin` and might have been missing in `C:\Windows\System32` before I installed the VS parts.

radekdoulik commented:
Sure, we shouldn't require the C/C++ runtime to be installed in the `System32`. I am not suggesting that. Instead we should let the dynamic loader to have it in the path.

So setting the environment like `PATH=<jre-bin-path>;%PATH%` should do it for us. Just tested that, renamed the System32 version and tried without and with the mentioned `PATH` setting. That works nicely.

Note to myself: it is possible to inspect dll dependencies issues on Windows with https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/gflags - that can be used to analyze issues during  runtime of the program running in the debugger.

Another tool to get whole dependency tree http://www.dependencywalker.com/ - it has its own issues though as well.

radekdoulik commented:
OK, finally got further with the crashes on Windows. We need to use different calling convention there. Will fix that in Java.Interop repository.

radekdoulik commented:
An update to share where we are now.

After fixing the calling conventions and more playing with mono dll's I am now able to get thru the generation of the methods to the point, where we move the types. We now crash when we try to write the updated assembly.

So it looks like our assembly unloading doesn't work on Windows.
```
Preparing marshal method assembly 'MobileApp2.Android-JniMarshalMethods'
Processing MobileApp2.Android.MainActivity type
Adding marshal method for Void OnCreate(Android.OS.Bundle)
Marshal method assembly 'MobileApp2.Android-JniMarshalMethods' created
Moving type MobileApp2.Android.MainActivity,C:\Users\rodo\source\repos\MobileApp2\MobileApp2\MobileApp2.Android\obj\Release\81\linksrc\MobileApp2.Android-JniMarshalMethods.dll to obj\Release\81\linksrc\MobileApp2.Android.dll
Moved method System.Void MobileApp2.Android.MainActivity/__<$>_jni_marshal_methods::.ctor()
Moved method System.Void MobileApp2.Android.MainActivity/__<$>_jni_marshal_methods::n_onCreate_Landroid_os_Bundle_(System.IntPtr,System.IntPtr,System.IntPtr)
Moved method System.Void MobileApp2.Android.MainActivity/__<$>_jni_marshal_methods::__RegisterNativeMembers(Java.Interop.JniNativeMethodRegistrationArguments)
Error: jnimarshalmethod-gen: Unable to process assembly 'obj\Release\81\linksrc\MobileApp2.Android.dll'
Win32 IO returned 1224. Path: C:\Users\rodo\source\repos\MobileApp2\MobileApp2\MobileApp2.Android\obj\Release\81\linksrc\MobileApp2.Android.dll
System.IO.IOException: Win32 IO returned 1224. Path: C:\Users\rodo\source\repos\MobileApp2\MobileApp2\MobileApp2.Android\obj\Release\81\linksrc\MobileApp2.Android.dll
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean anonymous, System.IO.FileOptions options) [0x00259] in D:\j\workspace\build-package-win-mono\2018-06\mcs\class\corlib\System.IO\FileStream.cs:274
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share) [0x00000] in D:\j\workspace\build-package-win-mono\2018-06\mcs\class\corlib\System.IO\FileStream.cs:91
  at (wrapper remoting-invoke-with-check) System.IO.FileStream..ctor(string,System.IO.FileMode,System.IO.FileAccess,System.IO.FileShare)
  at Mono.Cecil.ModuleDefinition.GetFileStream (System.String fileName, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share) [0x00008] in <df48fd6276ac478ab3758b42e3c25ba0>:0
  at Mono.Cecil.ModuleDefinition.Write (System.String fileName, Mono.Cecil.WriterParameters parameters) [0x00008] in <df48fd6276ac478ab3758b42e3c25ba0>:0
  at Mono.Cecil.AssemblyDefinition.Write (System.String fileName, Mono.Cecil.WriterParameters parameters) [0x00001] in <df48fd6276ac478ab3758b42e3c25ba0>:0
  at Xamarin.Android.Tools.JniMarshalMethodGenerator.TypeMover.Move () [0x0009d] in <d142345432a34dfb8d04171041bc6940>:0
  at Xamarin.Android.Tools.JniMarshalMethodGenerator.App.CreateMarshalMethodAssembly (System.String path) [0x00530] in <d142345432a34dfb8d04171041bc6940>:0
  at Xamarin.Android.Tools.JniMarshalMethodGenerator.App.ProcessAssemblies (System.Collections.Generic.List`1[T] assemblies) [0x0017d] in <d142345432a34dfb8d04171041bc6940>:0
```

I will try to find out whether that can be fixed. If not, we will just introduce a new option to the generator and save the assemblies to a different directory. Then let the task to move the files on Windows, after the generator run.

jonpryor commented:
We should add a new `jnimarshalmethod-gen.exe -o DIR` option, so that we can optionally create new files instead of updating the original files.

We do *not* want to use/rely on AppDomains, as this will complicate (far!) future .NET Core integrations, and Mono doesn't support `AssemblyLoadContext` yet.

radekdoulik commented:
An update: using `jnimarshalmethod-gen`'s `-o` option and using `Path.FullPath` in few places in `jnimarshalmethod-gen` I am now able to run on Windows.

jonpryor commented:
Closing as "mooted": while `jnimarshalmethod-gen` still (kinda) exists, it's now a `dotnet` app not a `monow` app; any and all integrations will need to be rethought from scratch.  There isn't much here that would be helpful.
